src/main/scala/beam/agentsim/infrastructure/ChargingNetworkManager.scala

```diff
@@ -0,0 +1,249 @@
+          scheduleTriggers :+ ScheduleTrigger(PlanningTimeOutTrigger(nextTick), self)
+        else {
+          // if we still have a BEV/PHEV that is connected to a charging point,
+          // we assume that they will charge until the end of the simulation and throwing events accordingly
+          val completeTriggers = scheduleTriggers ++ vehiclesToCharge.map {
+            case (_, cv) =>
+              ScheduleTrigger(
+                EndRefuelSessionTrigger(
+                  tick,
+                  cv.vehicle.getChargerConnectedTick(),
+                  cv.totalChargingSession.energy,
+                  cv.vehicle
+                ),
+                cv.agent
+              )
+          }
+          vehiclesToCharge.clear()
+          completeTriggers
+        }
+      )
+  }
+
+  private def replanHorizon(tick: Int): (Int, Map[Id[BeamVehicle], Double]) = {
+    val requiredPower = sitePowerManager.getPowerOverPlanningHorizon(vehicles)
+
+    val (bounds, nextTick) = if (gridConnectionEnabled) {
+      powerController.publishPowerOverPlanningHorizon(requiredPower, tick)
+      powerController.obtainPowerPhysicalBounds(tick)
+    } else {
+      powerController.defaultPowerPhysicalBounds(tick)
+    }
+    val requiredEnergyPerVehicle = sitePowerManager.replanHorizonAndGetChargingPlanPerVehicle(bounds, vehicles)
+
+    if (requiredEnergyPerVehicle.nonEmpty)
+      log.info("Required energy per vehicle: {}", requiredEnergyPerVehicle.mkString(","))
+
+    (nextTick, requiredEnergyPerVehicle)
+  }
+
+  private def endRefuelSessionTriggerMaybe(
+    vehicle: BeamVehicle,
+    tick: Int,
+    maxChargeDuration: Long,
+    currentSession: ChargingSession,
+    totalSession: ChargingSession
+  ): Option[EndRefuelSessionTrigger] = {
+    if (vehicle.primaryFuelLevelInJoules >= vehicle.beamVehicleType.primaryFuelCapacityInJoule) {
+      log.debug(
+        "Vehicle {} is fully charged. Scheduling EndRefuelSessionTrigger at {} with {} J delivered",
+        vehicle.id,
+        tick + currentSession.duration.toInt,
+        totalSession.energy
+      )
+      Some(
+        EndRefuelSessionTrigger(
+          tick + currentSession.duration.toInt,
+          vehicle.getChargerConnectedTick(),
+          totalSession.energy,
+          vehicle
+        )
+      )
+    } else if (currentSession.duration < maxChargeDuration) {
+      log.debug(
+        "Vehicle {} is charged by a short time: {}. Scheduling EndRefuelSessionTrigger at {} with {} J delivered",
+        vehicle.id,
+        currentSession.duration,
+        tick + currentSession.duration.toInt,
+        totalSession.energy
+      )
+      Some(
+        EndRefuelSessionTrigger(
+          tick + currentSession.duration.toInt,
+          vehicle.getChargerConnectedTick(),
+          totalSession.energy,
+          vehicle
+        )
+      )
+    } else {
+      log.debug(
+        "Ending refuel cycle for vehicle {}. Provided {} J. during {}",
```

Comment:
Is this log message correct? We are not ending the refuel since we are sending a None, correct?

Reply:
Yes, this is correct

Reply:
Thanks for the change to a boolean - makes more sense now :)